### PR TITLE
kconfig: Remove assignments to CONFIG_<arch> syms and hide them

### PR DESCRIPTION
**Pull request by ulfalizer**
(See
https://github.com/ulfalizer/zephyr/commits/hide-arch-syms-unsquashed
for a split-up version of this commit, which will make it easier to see
how stuff was done. This needs to go in as one commit though.)

All board defconfig files currently set the architecture in addition to
the board and the SoC, by setting e.g. `CONFIG_ARM=y`. This spams up
defconfig files.

`CONFIG_<arch>` symbols currently being set in configuration files also
means that they are configurable (can be changed in menuconfig and in
configuration files), even though changing the architecture won't work,
since other things get set from `-DBOARD=<board>`. Many boards also allow
changing the architecture symbols independently from the SoC symbols,
which doesn't make sense.

Get rid of all assignments to `CONFIG_<arch>` symbols and clean up the
relationships between symbols and the configuration interface, like
this:

1. Remove the `choice` with the `CONFIG_<arch>` symbols in `arch/Kconfig` and
   turn the `CONFIG_<arch>` symbols into invisible
   (promptless/nonconfigurable) symbols instead.

   Getting rid of the choice allows the symbols to be `select`ed (choice
   symbols don't support `select`).

2. Select the right `CONFIG_<arch>` symbol from the `SOC_SERIES_*` symbols.
   This makes sense since you know the architecture if you know the SoC.

   Put the select on the `SOC_*` symbol instead for boards that don't have
   a `SOC_SERIES_*`.

3. Remove all assignments to `CONFIG_<arch>` symbols. The assignments
   would generate errors now, since the symbols are promptless.

The change was done by grepping for assignments to `CONFIG_<arch>`
symbols, finding the `SOC_SERIES_*` (or `SOC_*`) symbol being set in the
same defconfig file, and putting a `select` on it instead.

This change is safer than it might seem re. outstanding PRs, because any
assignment to `CONFIG_<arch>` symbols generates an error now, making
outdated stuff easy to catch.

Also add a helpful warning:

```
kconfig/cmake: Check that one of the CONFIG_<arch> symbols is set

All SoCs must now 'select' one of the CONFIG_<arch> symbols. Add an
ARCH_IS_SET helper symbol that's selected by the arch symbols and
checked in CMake, printing a warning otherwise.

Might save people some time until they're used to the new scheme.
```

Additional refactoring for ARM, as a separate commit (to still give a good overview):

```
I verified that all board defconfigs (indirectly) set the CPU_CORTEX
symbol via either a SOC_SERIES_* symbol or via a SOC_* symbol, so the
'select ARM' can safely be moved to CPU_CORTEX, which unspams the SoC
Kconfigs.

Suggested by Kumar Gala.
```

### CODE REVIEW

**Comment by ulfalizer** [comment]
I'm planning to get rid of `CONFIG_SOC_*` and `CONFIG_BOARD_*` assignments from defconfig files later as well, but it needs to be split up a bit to be manageable, so starting with the `CONFIG_<arch>` symbols.

Nothing that can be derived from `-DBOARD=...` should need to be set.

**Comment by ulfalizer** [comment]
Some stuff can probably be cleaned up some more with `select` (which is fine for helper symbols), but this is a good start.

If one helper symbol logically implies another, then it's a good idea to select the second helper symbol from the first.

**Comment by ulfalizer** [comment]
This will always time out.

File: CMakeLists.txt [hunk 1]
```diff
@@ -142,6 +142,14 @@ else()
   assert(0 "Unreachable code. Expected optimization level to have been chosen. See Kconfig.zephyr")
 endif()
 
+if(NOT (CONFIG_ARC OR CONFIG_ARM OR CONFIG_X86 OR CONFIG_NIOS2 OR CONFIG_RISCV OR
```
Comment (erwango) [inline CMakeLists.txt]: I guess there is no way to make it more systematic, but this is likely to be missed in a next architecture porting.

**Review by erwango — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (aescolar) [inline CMakeLists.txt]: This would be quite annoying for out of tree architectures (like we have)

**Review by aescolar — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (SebastianBoe) [reply CMakeLists.txt]: Could the arch config's select HAS_ARCH?

Then we could just assert that HAS_ARCH is set.

**Review by SebastianBoe — APPROVED** [review]
This is a very welcome improvement, great work!

File: CMakeLists.txt (hunk 1 above)
Comment (aescolar) [reply CMakeLists.txt]: An intermediate symbol like that would certainly solve the problem

Comment (ulfalizer) [reply CMakeLists.txt]: Changed it to use a helper symbol.

Not sure how common this issue would be tbh, but might be nice for people who don't notice this change (though they'd get errors from assigning `CONFIG_<arch>` symbols).

**Review by aescolar — APPROVED** [review]

**Review by galak — DISMISSED** [review]
For most of the ARM cases we should move 'select ARM' to the `CPU_CORTEX_M`

**Comment by ulfalizer** [comment]
> For most of the ARM cases we should move 'select ARM' to the `CPU_CORTEX_M`

Could that be done as a separate change later? I didn't want to assume that people always associate the platform with just the CPU when I did the change, and this format makes it easy to verify too.

I think some other stuff CPU symbol stuff could be cleaned up too, but it gets too big if I piggyback it on this.

**Comment by ulfalizer** [comment]
Alright... looking at them some more, they just seem to be for ISA stuff, so it'd probably make sense to move some of the `select`s down to lower-level symbols.

There's other CPU stuff that could be cleaned up at the same time then though. Would be nice to have this in as a big but relatively easy-to-follow transformation.

**Comment by galak** [comment]
> Alright... looking at them some more, they just seem to be for ISA stuff, so it'd probably make sense to move some of the `select`s down to lower-level symbols.
> 
> There's other CPU stuff that could be cleaned up at the same time then though. Would be nice to have this in as a big but relatively easy-to-follow transformation.

Sure, I get that, but if we are shifting how ARCH is set, I think we should do it how we want at the end right away rather than making a transition.  I don't think anyone is going to care, plus it means less pain for any PRs in flight that add a new SoC.

**Comment by ulfalizer** [comment]
> Sure, I get that, but if we are shifting how ARCH is set, I think we should do it how we want at the end right away rather than making a transition. I don't think anyone is going to care, plus it means less pain for any PRs in flight that add a new SoC.

I'm worried that it'll delay the PR by several weeks with lots of unnecessary painful rebasing if I'll have to sort out all the nitty-gritty per-arch stuff in it too.

I never abandon stuff, and I rewrite things all the time. It's low-risk when it's me. :)

**Review by dcpleung — APPROVED** [review]

**Comment by galak** [comment]
> I'm worried that it'll delay the PR by several weeks with lots of unnecessary painful rebasing if I'll have to sort out all the nitty-gritty per-arch stuff in it too.
> 
> I never abandon stuff, and I rewrite things all the time. It's low-risk when it's me. :)

I don't think we need to get all the per-arch stuff set, just having `select ARM` on `config CORTEX` in `arch/arm/core/aarch32/cortex_m/Kconfig` and `arch/arm/core/aarch64/cortex_m/Kconfig` and similar for the other ARCH/CPU/CORE selections.

**Comment by ulfalizer** [comment]
> > I'm worried that it'll delay the PR by several weeks with lots of unnecessary painful rebasing if I'll have to sort out all the nitty-gritty per-arch stuff in it too.
> > I never abandon stuff, and I rewrite things all the time. It's low-risk when it's me. :)
> 
> I don't think we need to get all the per-arch stuff set, just having `select ARM` on `config CORTEX` in `arch/arm/core/aarch32/cortex_m/Kconfig` and `arch/arm/core/aarch64/cortex_m/Kconfig` and similar for the other ARCH/CPU/CORE selections.

Alright, I'll fix it up for the ARM stuff.

**Comment by ulfalizer** [comment]
Think I'll put it on `CPU_CORTEX` in `arch/arm/core/{aarch32,aarch64}/Kconfig`. Seems to be the lowest-level CPU symbols.

**Comment by ulfalizer** [comment]
@galak 
Added a separate commit to the PR that moves the `select ARM` to `CPU_CORTEX`.

**Comment by ulfalizer** [comment]
Hmm... those `depends on` causing the loop should probably go away.

**Comment by ulfalizer** [comment]
Looked into selecting `ARM`/`X86`/etc. from `CPU_*` symbols, but it gets messy, because each `arch/*/Kconfig` has this at the top level now:

```
menu "ARM/x86/etc. options"
	depends on ARM/X86/etc.

*All arch symbols*

endmenu
```

If anything within that menu selects `ARM`, there's a dependency loop, because then the symbol both depend on and selects `ARM`.

Fixing that might require some larger reorganization, like putting the CPU helper symbols outside the menu. Better to do that separately I think.

Another option would be to just drop the `depends on` from the menu, because it'll work out anyway with the Kconfig tree pruning. One drawback is that it means the full tree would make less sense, and it shows up e.g. in the generated documentation.

**Comment by ulfalizer** [comment]
Thinking it might be worth it to just start assuming pruning. Then there could just be

```
config ARM
	default y
```

at the top of `arch/arm/Kconfig` too.

**Comment by ulfalizer** [comment]
@galak 
ping

See https://github.com/zephyrproject-rtos/zephyr/pull/22566#issuecomment-583618120 for some investigation. Think it'd be too big a change for this PR.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
pushed a tweak for CI so we can get this tested.  marking DNM for right now.

**Comment by galak** [comment]
Merge this by hand as it passes the same tests as master, and isn't going to full pass.

**Pull request merged**